tools/yulPhaser/Population.cpp

```diff
@@ -0,0 +1,140 @@
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <tools/yulPhaser/Population.h>
+
+#include <tools/yulPhaser/Program.h>
+
+#include <algorithm>
+#include <cassert>
+#include <iostream>
+#include <numeric>
+
+using namespace std;
+using namespace solidity::langutil;
+
+namespace solidity::phaser
+{
+
+size_t Population::measureFitness(Chromosome const& _chromosome, CharStream& _sourceCode)
+{
+	auto program = Program::load(_sourceCode);
+	program.optimise(_chromosome.optimisationSteps());
+	return program.codeSize();
+}
+
+ostream& operator<<(ostream& _stream, Individual const& _individual)
+{
+	_stream << "Fitness: ";
+	if (_individual.fitness.has_value())
+		_stream << _individual.fitness.value();
+	else
+		_stream << "<NONE>";
+	_stream << ", optimisations: " << _individual.chromosome;
+
+	return _stream;
+}
+
+Population::Population(CharStream _sourceCode, vector<Chromosome> const& _chromosomes):
+	m_sourceCode{std::move(_sourceCode)}
+{
+	for (auto const& chromosome: _chromosomes)
+		m_individuals.push_back(Individual{chromosome, nullopt});
+}
+
+Population Population::makeRandom(CharStream _sourceCode, size_t _size)
+{
+	vector<Individual> individuals;
+	for (size_t i = 0; i < _size; ++i)
+		individuals.push_back(Individual{
+			Chromosome::makeRandom(randomChromosomeLength()),
+			nullopt
+		});
+
+	return Population(move(_sourceCode), individuals);
+}
+
+void Population::run(optional<size_t> _numRounds, std::ostream& _outputStream)
+{
+	assert(_numRounds == nullopt || _numRounds >= 0);
+
+	doEvaluation();
+	for (size_t round = 0; !_numRounds.has_value() || round < _numRounds.value(); ++round)
+	{
+		doMutation();
+		doSelection();
+		doEvaluation();
+
+		_outputStream << "---------- ROUND " << round << " ----------" << endl;
+		_outputStream << *this;
+	}
+}
+
+ostream& operator<<(ostream& _stream, Population const& _population)
+{
+	_stream << "Stream name: " << _population.m_sourceCode.name() << endl;
+
+	auto individual = _population.m_individuals.begin();
+	for (; individual != _population.m_individuals.end(); ++individual)
+		_stream << *individual << endl;
+
+	return _stream;
+}
+
+void Population::doMutation()
+{
+	// TODO: Implement mutation and crossover
+}
+
+void Population::doEvaluation()
+{
+	for (auto& individual: m_individuals)
+		if (!individual.fitness.has_value())
+			individual.fitness = measureFitness(individual.chromosome, m_sourceCode);
+}
+
+void Population::doSelection()
+{
+	assert(all_of(m_individuals.begin(), m_individuals.end(), [](auto& i){ return i.fitness.has_value(); }));
+
+	sort(
+		m_individuals.begin(),
+		m_individuals.end(),
+		[](auto const& a, auto const& b){ return a.fitness.value() < b.fitness.value(); }
+	);
+
+	randomizeWorstChromosomes(m_individuals, m_individuals.size() / 2);
```

Comment:
Maybe it is more flexible to do:

```
size_t desiredSize = m_individuals.size();
m_individuals.resize(desiredSize / 2);
refillPool(m_individuals, desiredSize);
```

And then refill does:

```
while (m_individuals.size() < _desiredSize)
  m_individuals.emplace_back(generateNewIndividual(m_individuals));
```

Reply:
Yeah. That's a good idea for refactoring the current design but since I'm changing that design in the next PR anyway, I'll leave it as is here.